content/posts/2025-gsoc-clang-doc.md

```diff
@@ -0,0 +1,207 @@
+---
+author: "Erick Velez (evelez7)"
+date: "2025-08-28"
+tags: ["GSoC", "clang-doc", "clang-tools-extra", "documentation"]
+title: "GSoC 2025: Improving Core Clang-Doc Functionality"
+---
+
+I was selected as a contributor for GSoC 2025 under the project "Improving Core Clang-Doc Functionality" for LLVM.
+My mentors for the project were Paul Kirth and Petr Hosek.
+
+Clang-Doc is a tool in clang-tools-extra that generates documentation from Clang's AST.
+Clang-Doc can output documentation in Markdown, HTML, YAML, and JSON.
+The project started in 2018 but major development eventually slowed.
+Recently, there have been efforts to get it back on track.
+
+This year, the [GSoC project idea](https://discourse.llvm.org/t/improve-documentation-parsing-in-clang/84513) had a simple premise: improve core functionality.
+
+## The Issues
+
+The project idea proposed three main areas of focus to improve documentation quality.
+
+1. C++ support
+2. Doxygen comments
+3. Markdown support
+
+First, not all C++ constructs were supported, like friends or concepts.
+Not supporting core C++ constructs in C++ documentation is not good.
+Second, it's important that Doxygen command support is robust and that we can support as many as possible.
+Third and last, having Markdown available to developers for documentation would be useful.
+Markdown provides the power of expression in an area that is technically dense.
+It can be used to highlight critical information and warnings.
+
+### The Architecture
+
+Here's a quick overview on Clang-Doc's architecture, which follows a map-reduce pattern:
+
+1. Visit source declarations via Clang's ASTVisitor.
+2. Serialize relevant source information into an Info (Clang-Doc's main data entity).
+3. Once all source declarations are serialized, write them into bitcode, reduce, and read the reduced Infos.
+4. Serialize Infos into the desired format.
+
+It seems fairly straightforward, but the architecture had a critical flaw.
+If a new C++ construct needed to be supported, it would be visited and serialized, but then support would have to be added to each backend individually.
+If you wanted to serialize something in YAML, you'd have to implement the Markdown logic separately.
+This placed a very high maintenance cost for extending basic functionality, even if you just wanted to add something simple.
+It also easily led to generator disparity; a construct might be serialized in YAML, but not in Markdown.
+
+Testing was also in an awkward spot because it was unclear what format would be used to verify if the documentation output was acceptable.
+YAML was the initial candidate for this, but my mentors had started to consider JSON instead.
+Feature parity was far apart; some backends were tested for certain attributes that others didn't have.
```

Comment:
These all seem like disjoint ideas. maybe tie them together a bit more, or just expand this section slightly?